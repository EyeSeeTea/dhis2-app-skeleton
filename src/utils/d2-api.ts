import { D2Api } from "../types/d2-api";

interface Instance {
    url: string;
    username?: string;
    password?: string;
}

export function getD2APiFromInstance(instance: Instance) {
    if (instance.username && instance.password) {
        return new D2Api({
            baseUrl: instance.url,
            auth: { username: instance.username, password: instance.password },
            backend: "fetch",
        });
    }

    return new D2Api({ baseUrl: instance.url, backend: "fetch" });
}
