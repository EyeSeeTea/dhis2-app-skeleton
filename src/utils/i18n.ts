import { getModuleForNamespace } from "./i18n-typed";

export default getModuleForNamespace("dhis2-skeleton-app");
