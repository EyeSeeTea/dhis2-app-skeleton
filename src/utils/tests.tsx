import { render, RenderResult } from "@testing-library/react";
import { SnackbarProvider } from "@eyeseetea/d2-ui-components";
import { ReactNode } from "react";
import { AppContext, AppContextState } from "$/webapp/contexts/app-context";
import { getTestCompositionRoot } from "$/CompositionRoot";
import { createAdminUser } from "$/domain/entities/__tests__/userFixtures";

export function getTestContext() {
    const context: AppContextState = {
        currentUser: createAdminUser(),
        compositionRoot: getTestCompositionRoot(),
    };

    return context;
}

export function getReactComponent(children: ReactNode): RenderResult {
    const context = getTestContext();

    return render(
        <AppContext.Provider value={context}>
            <SnackbarProvider>{children}</SnackbarProvider>
        </AppContext.Provider>
    );
}
