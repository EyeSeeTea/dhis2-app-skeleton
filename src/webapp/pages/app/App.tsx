import styled from "styled-components";
import { HeaderBar } from "@dhis2/ui";
import { SnackbarProvider } from "@eyeseetea/d2-ui-components";
import { Feedback } from "@eyeseetea/feedback-component";
import { MuiThemeProvider } from "@material-ui/core/styles";
//@ts-ignore
import OldMuiThemeProvider from "material-ui/styles/MuiThemeProvider";
import React, { useEffect, useState } from "react";
import { appConfig } from "$/app-config";
import { CompositionRoot } from "$/CompositionRoot";
import Share from "$/webapp/components/share/Share";
import { AppContext, AppContextState } from "$/webapp/contexts/app-context";
import { Router } from "$/webapp/pages/Router";
import "./App.css";
import muiThemeLegacy from "./themes/dhis2-legacy.theme";
import { muiTheme } from "./themes/dhis2.theme";

export interface AppProps {
    compositionRoot: CompositionRoot;
}

function App(props: AppProps) {
    const { compositionRoot } = props;
    const [showShareButton, setShowShareButton] = useState(false);
    const [loading, setLoading] = useState(true);
    const [appContext, setAppContext] = useState<AppContextState | null>(null);

    useEffect(() => {
        async function setup() {
            const isShareButtonVisible = appConfig.appearance.showShareButton;
            const currentUser = await compositionRoot.users.getCurrent.execute().toPromise();
            if (!currentUser) throw new Error("User not logged in");

            setAppContext({ currentUser, compositionRoot });
            setShowShareButton(isShareButtonVisible);
            setLoading(false);
        }
        setup();
    }, [compositionRoot]);

    if (loading) return null;

    return (
        <MuiThemeProvider theme={muiTheme}>
            <OldMuiThemeProvider muiTheme={muiThemeLegacy}>
                <SnackbarProvider>
                    <StyledHeaderBar appName="Skeleton App" />

                    {appConfig.feedback && appContext && (
                        <Feedback
                            options={appConfig.feedback}
                            username={appContext.currentUser.username}
                        />
                    )}

                    <div id="app" className="content">
                        <AppContext.Provider value={appContext}>
                            <Router />
                        </AppContext.Provider>
                    </div>

                    <Share visible={showShareButton} />
                </SnackbarProvider>
            </OldMuiThemeProvider>
        </MuiThemeProvider>
    );
}

const StyledHeaderBar = styled(HeaderBar)`
    div:first-of-type {
        box-sizing: border-box;
    }
`;

export default React.memo(App);
