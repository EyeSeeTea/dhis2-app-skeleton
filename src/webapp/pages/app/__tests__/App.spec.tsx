import { fireEvent, render } from "@testing-library/react";

import App from "$/webapp/pages/app/App";
import { getTestContext } from "$/utils/tests";
import { Provider } from "@dhis2/app-runtime";

describe("App", () => {
    it("renders the feedback component", async () => {
        const view = getView();

        expect(await view.findByText("Send feedback")).toBeInTheDocument();
    });

    it("navigates to page", async () => {
        const view = getView();

        fireEvent.click(await view.findByText("John"));

        expect(await view.findByText("Hello John")).toBeInTheDocument();
        expect(view.asFragment()).toMatchSnapshot();
    });
});

function getView() {
    const { compositionRoot } = getTestContext();
    return render(
        <Provider config={{ baseUrl: "http://localhost:8080", apiVersion: 30 }}>
            <App compositionRoot={compositionRoot} />
        </Provider>
    );
}
