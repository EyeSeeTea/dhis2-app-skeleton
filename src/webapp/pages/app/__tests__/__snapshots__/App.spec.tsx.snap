// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > navigates to page 1`] = `
<DocumentFragment>
  <header
    class="jsx-2837859311 sc-dhKdPU klBWsW"
  />
  <div
    class="Component-root-4 Component-root-5"
  >
    <div
      style="padding: 2px 10px 2px 10px; font-size: 12px; cursor: pointer; color: rgb(255, 255, 255); background-color: rgb(255, 152, 0); font-family: \\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif; font-weight: 500; line-height: 1.75; letter-spacing: 0.02857em; text-transform: uppercase;"
    >
      Send feedback 
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
        focusable="false"
        style="padding-top: 2px; margin-left: 3px;"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 2H4c-1.1 0-1.99.9-1.99 2L2 22l4-4h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm-7 12h-2v-2h2v2zm0-4h-2V6h2v4z"
        />
      </svg>
    </div>
  </div>
  <div
    class="content"
    id="app"
  >
    <div>
      <button
        aria-label="Back"
        class="MuiButtonBase-root MuiIconButton-root sc-gEvDqW sc-eqUzNf hwoczq haqPoX MuiIconButton-colorSecondary"
        data-test="page-header-back"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <span
            aria-hidden="true"
            class="material-icons MuiIcon-root MuiIcon-colorPrimary"
          >
            arrow_back
          </span>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <h5
        class="MuiTypography-root sc-aYaIB hZQTlr MuiTypography-h5 MuiTypography-gutterBottom"
        data-test="page-header-title"
      >
        Detail page
      </h5>
    </div>
    <h2
      class="sc-fqkwJk lcQFEV"
    >
      Hello John
    </h2>
  </div>
  <div>
    <div
      style="bottom: -3px; right: 100px; position: fixed; z-index: 10002;"
    >
      <button
        style="text-shadow: none; background-color: rgb(255, 152, 0); color: white; width: 65px; height: 38.5px; cursor: pointer; border: 1px solid rgba(0, 0, 0, 0.1); border-radius: 2px; background-clip: padding-box; box-shadow: 0 4px 16px rgba(0, 0, 0, 0.2);"
      >
        <i
          class="fa fa-share icon-xlarge"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
